lib/action.ml

```diff
@@ -171,24 +170,33 @@ module Action (Github_api : Api.Github) (Slack_api : Api.Slack) = struct
       in
       Lwt.return (direct_message @ chans)
     in
-    if Context.is_pipeline_allowed ctx repo.url ~pipeline then begin
-      let repo_state = State.find_or_add_repo ctx.state repo.url in
-      match Rule.Status.match_rules ~rules n with
-      | Some (Ignore, _, _) | None -> Lwt.return []
-      | Some (Allow, notify_channels, notify_dm) -> action_on_match n.branches ~notify_channels ~notify_dm
-      | Some (Allow_once, notify_channels, notify_dm) ->
-      match StringMap.find_opt pipeline repo_state.pipeline_statuses with
-      | Some branch_statuses ->
-        let has_same_status_as_prev (branch : branch) =
-          match StringMap.find_opt branch.name branch_statuses with
-          | Some state when state = current_status -> true
-          | _ -> false
-        in
-        let branches = List.filter (Fun.negate has_same_status_as_prev) n.branches in
-        action_on_match branches ~notify_channels ~notify_dm
-      | None -> action_on_match n.branches ~notify_channels ~notify_dm
-    end
-    else Lwt.return []
+    let%lwt recipients =
+      if Context.is_pipeline_allowed ctx repo.url ~pipeline then begin
+        let repo_state = State.find_or_add_repo ctx.state repo.url in
+        match Rule.Status.match_rules ~rules n with
+        | Some (Ignore, _, _) | None -> Lwt.return []
+        | Some (Allow, notify_channels, notify_dm) -> action_on_match n.branches ~notify_channels ~notify_dm
+        | Some (Allow_once, notify_channels, notify_dm) ->
+        match StringMap.find_opt pipeline repo_state.pipeline_statuses with
+        | Some branch_statuses ->
+          let has_same_status_as_prev (branch : branch) =
+            match StringMap.find_opt branch.name branch_statuses with
+            | Some state when state = current_status -> true
+            | _ -> false
+          in
+          let branches = List.filter (Fun.negate has_same_status_as_prev) n.branches in
+          let notify_dm =
+            match branches with
+            | [] -> false
+            | _ -> notify_dm
+          in
+          action_on_match branches ~notify_channels ~notify_dm
+        | None -> action_on_match n.branches ~notify_channels ~notify_dm
+      end
+      else Lwt.return []
+    in
+    State.set_repo_pipeline_status ctx.state repo.url ~pipeline ~branches:n.branches ~status:current_status;
```

Comment:
I think that it is necessary to move the set state here so that it happens every time. Otherwise we can have a branch which is marked as failure and then will never see its status being changed again?